substrate/in-instructions/primitives/src/lib.rs

```diff
@@ -0,0 +1,47 @@
+#![cfg_attr(docsrs, feature(doc_cfg))]
+#![cfg_attr(docsrs, feature(doc_auto_cfg))]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use scale::{Encode, Decode, MaxEncodedLen};
+use scale_info::TypeInfo;
+
+#[cfg(feature = "std")]
+use serde::{Serialize, Deserialize};
+
+use sp_core::{ConstU32, bounded::BoundedVec};
+
+// Monero, our current longest address candidate, has a longest address of featured with payment ID
+// 1 (enum) + 1 (flags) + 64 (two keys) + 8 (payment ID) = 74
+pub const MAX_ADDRESS_LEN: u32 = 74;
+// Should be enough for a Uniswap v3 call
+pub const MAX_DATA_LEN: u32 = 512;
+
+#[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, MaxEncodedLen, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct ExternalAddress(BoundedVec<u8, ConstU32<{ MAX_ADDRESS_LEN }>>);
+impl ExternalAddress {
+  #[cfg(feature = "std")]
+  pub fn new(address: Vec<u8>) -> Result<ExternalAddress, &'static str> {
+    Ok(ExternalAddress(address.try_into().map_err(|_| "address length exceeds {MAX_ADDRESS_LEN}")?))
+  }
+
+  pub fn address(&self) -> &[u8] {
+    self.0.as_ref()
+  }
+
+  #[cfg(feature = "std")]
+  pub fn consume(self) -> Vec<u8> {
+    self.0.into_inner()
+  }
+}
+
+// Not "in" as "in" is a keyword
+mod incoming;
+pub use incoming::*;
+
+// Not "out" to match in
+mod outgoing;
+pub use outgoing::*;
+
+mod shorthand;
```

Comment:
txencoded?

Reply:
1) I'd rather not prefix everything with "tx"
2) No. All of these can be encoded on their own right. We just always use Shorthand for external networks representations.

Reply:
To understand shorthand someone has to look up the definition in our context. This in contrast to the transaction memo TC uses, TC wins on ease of understanding. We should add something to shorthand to make it more clear.

Reply:
I feel every protocol has to be looked up to be properly understood. "Shorthand" is a long-standing term for a quick notation which expands to something. I'd argue "memo" is ambiguous since it suggests a message, not an instruction. THORChain isn't a chat app.

So yes, you have to look up what Shorthand expands to. That's called reading the docs to see how the protocol works. We can't handhold people through every single step of the protocol by how we name things alone. Even if it "InInstructionShorthand", they'd still have to lookup the options/fields. Even if it was "txencoded", they'd have to look up what the encoding format is and what the TX is.

"txencoded" is wrong for suggesting it's THE encoding when it isn't. These also aren't TXs. I'd like to close this and move on unless you can suggest a better name.

Reply:
This is also an internal file name. It just has its members exported as `Shorthand`.